Clustering.Rmd

```diff
@@ -0,0 +1,621 @@
+sample = sample.split(quant_data, SplitRatio = .7)
+
+train = subset(quant_data, sample == TRUE)
+test  = subset(quant_data, sample == FALSE)
+
+trainS=scale(train)
+testS=scale(test,center = colMeans(train),scale = apply(train,2,sd))
+
+```
+
+Now let us train the model again
+```{r}
+res.2 <- as.data.frame(matrix(nrow=8,ncol=3))
+colnames(res.2) <- c("k", "train.VAF","holdout.VAF")
+rownames(res.2)<-c()
+n=1
+
+for (k in seq(2,5)){
+        res.2[n,"k"]=k
+        km.train <- kmeans(trainS, algorithm="Lloyd",centers = k, nstart = 50,iter.max=100)
+        res.2[n,"train.VAF"] <- 1 - km.train$tot.withinss / km.train$totss
+        km.holdout <- kmeans(testS, centers = km.train$centers,algorithm="Lloyd", nstart = 1,iter.max=100)
+        res.2[n,"holdout.VAF"] <- 1 - km.holdout$tot.withinss / km.holdout$totss
+        n=n+1
+}
+res.2
+```
+
+Let us select k=6
+```{r}
+km.train.6 <- kmeans(trainS, algorithm="Lloyd",centers = 6, nstart = 50, iter.max=1000)
+km.train.6.VAF=1 - km.train.6$tot.withinss / km.train.6$totss
+km.train.6.VAF
+```
+
+Let us plot to analyze the centers
+```{r}
+library(reshape2)
+library(ggpubr)
+trainMatrix=as.data.frame(matrix(nrow=5, ncol=7))
+
+trainMatrix[,2:7]=as.data.frame(t(km.train.6$centers))
+trainMatrix[,1]=c("duration_seconds","miles","tip","extra_charges","total_no_tip")
+colnames(trainMatrix) = c("ClusterName","ClusterOne","ClusterTwo","ClusterThree","ClusterFour","ClusterFive","ClusterSix")
+
+trainMatrix<-melt(trainMatrix,id.vars="ClusterName")
+
+ggplot(trainMatrix, aes(x=ClusterName,y=value, colour=variable, group=variable )) + geom_line()+geom_point(size = 3)+ggtitle("Training data centers 6 clusters Kmeans with selected variables")+ theme(plot.title = element_text(hjust=0.5, size=15),axis.text.x = element_text(angle=55, hjust=1))
+```
+Let us plot without cluster three
+```{r}
+trainMatrix=as.data.frame(matrix(nrow=5, ncol=6))
+
+trainMatrix[,2:6]=as.data.frame(t(km.train.6$centers[c(1,2,4,5,6),]))
+trainMatrix[,1]=c("duration_seconds","miles","tip","extra_charges","total_no_tip")
+colnames(trainMatrix) = c("ClusterName","ClusterOne","ClusterTwo","ClusterFour","ClusterFive","ClusterSix")
+trainMatrix<-melt(trainMatrix,id.vars="ClusterName")
+
+ggplot(trainMatrix, aes(x=ClusterName,y=value, colour=variable, group=variable )) + geom_line()+geom_point(size = 3)+ggtitle("Training data centers 6 clusters Kmeans with fare related variables")+ theme(plot.title = element_text(hjust=0.5, size=15),axis.text.x = element_text(angle=55, hjust=1))
+```
+
+Let us try to select **traffic** related variables
+```{r}
+head(cleaned_data)
+```
+
+```{r}
+require(caTools)
+set.seed(232323)
+
+#select only numeric variables
+quant_data=cleaned_data[,c("speed_pickup_start","bus_count_pickup_start","gps_pings_pickup_start","speed_dropoff_end","bus_count_dropoff_end","gps_pings_dropoff_end")]
+sample = sample.split(quant_data, SplitRatio = .7)
+
+train = subset(quant_data, sample == TRUE)
```

Comment:
Also, is there a specific reason you re-sampled train and test?

Reply:
It's just easier for me to run. I don't have to run from top to bottom if I need some specific clustering